### PR TITLE
Adding precaching support to speed up linear evals

### LINKED ISSUE #19: Add pre-caching for linear eval
- [x] Save features to SLURM_TMPDIR using ffcv dataloader
- [x] Use simple dataloader to train linear classifier from pre-cached features
- [x] Test accuracy numbers wrt no precaching

### PR DESCRIPTION
**Pull request by arnaghosh**
**TL;DR: Precaching support added, steps added in [verify_barlowTwins.sh](https://github.com/kumarkrishna/fastssl/blob/arna/precaching/sbatch_sh_files/verify_barlowTwins.sh) demonstrating how to precache+linear eval.**

**Summary of changes:**

1. *cc_precache.yaml*: Added new yaml config file that define the arguments while running precaching

2. *cc_classifier.yaml*: Added a new eval argument `use_precache` to switch between precached linear eval and no precaching

3. *simple_dataloaders.py*: Added a file to define simple pytorch dataloader for loading a npy file from disk and yielding batches of activations for linear eval.

4. *linear.py*: Changed the definition of backbone to Identity when precached features are being used.

5. *train.py*: 
   * Added precache argument to training args for the usecase when no training will be done
   * Added use_precache argument to eval args for the usecase when precached features are to be used for eval
   * Added function `precache_outputs` that saves the model outputs of train and test loaders to $SLURM_TMPDIR as npy file (dict object)
   * Added function `search_precache_file` to search for a precached file when running linear eval (with `use_precache=True`), default to no precaching (i.e. changing `eval.use_precache` to False) if precached file not found.
   * Changed `build_dataloaders` function to return `SimpleDataloader` when training.train_dataset points to a npy file
   * Changed `build_model` to set `model.bkey` to an empty string when precaching is being done.
   * Added `model.backbone.eval()` in `train_step` function when doing linear eval to ensure that the backbone is in eval model, i.e. frozen params as well as statistics. 

**Test plan:**
- Ran precaching and checked if files are saved properly.
- Compare features with and without precaching, i.e. check that at every epoch of non-precached linear eval, the features that were used as input to train the linear layer were the same as those saved during precaching.
- Compared performance of linear eval with and without precaching.
- There is a slight difference in performance numbers (tried for 20 epochs of training). However, the differences become very small (~0.1%) when the ordering of the data is set to be SEQUENTIAL instead of RANDOM (i.e. no shuffling). This indicates that the performance difference originates from randomness in batch sampling when using multiple workers (ffcv issue)

This PR should close #19 and perhaps also #2.  

**EDIT:** Adding some benchmarks on speedup from precaching. Ran experiments on A100 MIG (Mila cluster's `2g.20gb` node) for *20 epochs* of linear readout training.

| Precaching  | Time taken | Accuracy |
| -----------  | ----------- | --------- |
| No   | 216.6 s  | 81.29 % |
| Yes  | 32.25 s*  | 81.07 % |

*Time taken to precache features/embeddings: ~22.5 s, hence total time is (22.5 + 32.25) s = 54.75 s for 20 epochs of linear readout training.

### CODE REVIEW

File: fastssl/data/simple_dataloaders.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import numpy as np
+from torch.utils.data import Dataset, DataLoader
+
+class SimpleDataset(Dataset):
+    def __init__(self,
+                 data_labels_dict,
+                 X_key='activations',
```
Comment (kumarkrishna) [inline fastssl/data/simple_dataloaders.py:7]: nit: let's use small case for all vars

File: fastssl/data/simple_dataloaders.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        self.X = data_labels_dict[X_key]
+        self.y = data_labels_dict[y_key]
+
+    def __len__(self):
+        return len(self.X)
+
+    def __getitem__(self,idx):
+        return self.X[idx], self.y[idx]
+    
+def SimpleDataloader(fname_train,
```
Comment (kumarkrishna) [inline fastssl/data/simple_dataloaders.py:18]: nit: snake_case for funcs

File: fastssl/data/simple_dataloaders.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                     fname_test,
+                     splits=['train','test'],
+                     batch_size=512,
+                     num_workers=2):
+    assert fname_train==fname_test, "Precaching train/test features should be stored in the same file!"
```
Comment (kumarkrishna) [inline fastssl/data/simple_dataloaders.py:23]: should we pass a single file in this case? or do you want to support diff files in the future?

File: fastssl/data/simple_dataloaders.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    data_from_file = np.load(fname_train,allow_pickle=True).item()
+    loaders = {}
+    for split in splits:
+        dataset = SimpleDataset(data_labels_dict=data_from_file[split])
+        loaders[split] = DataLoader(
+            dataset, 
+            batch_size=batch_size,
+            shuffle=True,
```
Comment (kumarkrishna) [inline fastssl/data/simple_dataloaders.py:31]: recommend `shuffle=False` for test-split

File: fastssl/models/linear.py [hunk 5]
```diff
@@ -57,6 +64,9 @@ def forward(self, x):
             feats = self.backbone.proj(x)
         elif 'feat' in self.bkey:
             feats = self.backbone(x)
+        elif len(self.bkey)==0:
```
Comment (kumarkrishna) [inline fastssl/models/linear.py:67]: recommend:
```
x = self.backbone(x)
if 'proj' in self.bkey:
   x = torch.flatten(x, start_dim=1)
```

File: fastssl/models/linear.py [hunk 6]
```diff
@@ -57,6 +64,9 @@ def forward(self, x):
...
+            # no backbone, i.e. Identity case
+            feats = self.backbone(x)
         feats = torch.flatten(feats, start_dim=1)
```
Comment (kumarkrishna) [inline fastssl/models/linear.py:70]: q: are we flattening twice?

**Review by kumarkrishna — CHANGES_REQUESTED** [review]
looks good! added some minor comments. 
Is adding a quick benchmark on speedups when using caching features easy?

File: fastssl/data/simple_dataloaders.py (hunk 3 above)
Comment (arnaghosh) [reply fastssl/data/simple_dataloaders.py:23]: Discussed and ignored this suggestion.

File: fastssl/models/linear.py (hunk 6 above)
Comment (arnaghosh) [reply fastssl/models/linear.py]: Discussed that this is not an issue.

**Review by arnaghosh — COMMENTED** [review]
Made changes as requested and discussed :)

**Review by kumarkrishna — APPROVED** [review]
lgtm! wondering where we're logging the benchmark. that can be addressed independently in a different pr

**Comment by arnaghosh** [comment]
I kept the benchmark table in the PR description. But we can also keep a separate README for these numbers?
